internal/external-dns/plan/plan.go

```diff
@@ -200,6 +206,60 @@ func (p *Plan) Error() error {
 	return errors.Join(p.Errors...)
 }
 
+func CleanLabelFromEndpoint(label string, endpoint *endpoint.Endpoint) *endpoint.Endpoint {
+	if endpoint.Labels == nil {
+		return endpoint
+	}
+
+	delete(endpoint.Labels, label)
+	return endpoint
+}
+
+func CleanLabelFromEndpoints(label string, endpoints []*endpoint.Endpoint) []*endpoint.Endpoint {
+	for _, e := range endpoints {
+		CleanLabelFromEndpoint(label, e)
+	}
+	return endpoints
+}
+func CleanLabelsFromEndpoints(labels []string, endpoints []*endpoint.Endpoint) []*endpoint.Endpoint {
+	for _, l := range labels {
+		CleanLabelFromEndpoints(l, endpoints)
+	}
+
+	return endpoints
+}
+
+func (p *Plan) processStopLabel() {
+	if p.RootHost == nil {
+		return
+	}
+	desiredTree := common.MakeTreeFromEndpoints(*p.RootHost, p.Desired)
+	currentTree := common.MakeTreeFromEndpoints(*p.RootHost, p.Current)
+
+	// bad trees, return
+	if len(desiredTree.Children) == 0 || len(currentTree.Children) == 0 {
+		return
+	}
+
+	// copy soft_delete and stop_soft_delete labels from desired to current tree for propagation
+	common.CopyLabel(SoftDeleteLabel, desiredTree, currentTree)
+	common.CopyLabel(StopSoftDeleteLabel, desiredTree, currentTree)
+
+	// propagate label in desired tree ignoring stop labels
+	common.PropagateLabel(desiredTree, SoftDeleteLabel, "true")
+
+	// propagate the stop label in the current tree
+	common.PropagateStoppableLabel(currentTree, SoftDeleteLabel, "true", StopSoftDeleteLabel)
+
+	// write everything back to the plan
+	newDesired := []*endpoint.Endpoint{}
+	newCurrent := []*endpoint.Endpoint{}
+	common.ToEndpoints(desiredTree, &newDesired)
+	common.ToEndpoints(currentTree, &newCurrent)
+	p.Desired = common.MergeEndpoints(newDesired, p.Desired)
+	p.Current = common.MergeEndpoints(newCurrent, p.Current)
+}
```

Comment:
Does this have to create entirely new slices of endpoints? Can it not just label the endpoints in memory or is it because its going through this tree code, does it not just maintain a pointer to the endpoint structs in the tree?

Reply:
it does maintain a handle to the endpoints, and I originally did as you say, however the problem caused here was that the tree ignores any endpoints that aren't able to be traced back to the root, and in some tests those records are expected to be preserved. I changed this code to get those tests to pass again.